pybind/snap_schedule: add initial interface

jan--f opened the pull request:
This is an RFC PR for the proposed interface for the cephfs snapshot schedule module.

This is base on a CDM discussion from June 5th (https://youtu.be/EG-3rcvqvGI?list=PLrBUGiINAakNbcSOvOM0IJHqqv5dzusZ6&t=1812), more details here https://pad.ceph.com/p/cephfs-snap-mirroring.

A `snap_schedule` is simply a multiplier and a time spec (m[inute], h[our], ...) to snapshot a path every so often. The start of this period is the time `set` was called. A (not yet added) `start` argument would allow more control over the timing. Imho this is preferable over adding a start time to the `snap_schedule` spec (something like `24h@<start_time>`).

A `prune_schedule` is optional. By default a pruning is attempted after every scheduled snapshot. If specified is has them same specification as `snap_schedule`.

One thing that is not included yet would be to allow multiple snap (or prune) schedules per path. This would allow adding complex snapshot schedules without complicating the `snap_schedule` specification. I think this can easily be added to the interface at a later stage (index the snapshot schedules per path and add an optional index arg the the `get`, `set` and `rm` calls).

batrick commented:
https://tracker.ceph.com/issues/41072

jan--f commented:
cc @oliveiradan

jan--f commented:
Should we allow setting snapshot schedules for not (yet) existing directories? This might be handy for use cases where provisioning scripts want to create snapshot schedules but not the full directory tree.
What goes hand in hand with this is that we don't necessarily have to clean up snapshot schedules when a directory is deleted.

l-mb commented:
Should we able to set a jitter period to avoid multiple snaps triggering at exactly the same time?

liewegas commented on src/pybind/mgr/snap_schedule/module.py:
```diff
@@ -0,0 +1,155 @@
+        raise NotImplementedError
+
+    @CLIReadCommand('fs snap-schedule get',
+                    'name=path,type=CephString '
```
Is this path relative to the vol/subvol?

jan--f replied:
Yes, if a subvolume is passed. Iiuc that is how [sub]volumes are currently implemented. I suppose they could also imply a specific fs instance to be used.

liewegas commented:
I don't really understand why pruning has a schedule at all.  It seems like each snap schedule item would have some retention property (how many of these), and that determines what gets pruned.  Whether a prune happens 2 seconds or 2 hours after the count associated with a schedule is changed isn't something a user should ever have to think about; normally the prune would happen after a new snap is taken anyway.

Note that having a snap schedule have a single period means that if you want (say) daily and weekly snapshots, then you'll have 2 snapshots every 7th day that sort of overlap.  This is why the original proposal was trying to specify a more complicated retention policy that would thin out the snapshot history over time as part of the same schedule.  I still think this would be a nicer experience, as long as we can come up with a way of describing that policy in an understandable way.

jan--f commented:
> I don't really understand why pruning has a schedule at all. It seems like each snap schedule item would have some retention property (how many of these), and that determines what gets pruned. Whether a prune happens 2 seconds or 2 hours after the count associated with a schedule is changed isn't something a user should ever have to think about; normally the prune would happen after a new snap is taken anyway.

Right. That would of course be the default (the prune_schedule is optional). I thought that might be something user would want to configure, like "only prune at night" or so. But not feeling strongy about this...probably best we leave that out for now.

jan--f commented:
> Note that having a snap schedule have a single period means that if you want (say) daily and weekly snapshots, then you'll have 2 snapshots every 7th day that sort of overlap. This is why the original proposal was trying to specify a more complicated retention policy that would thin out the snapshot history over time as part of the same schedule. I still think this would be a nicer experience, as long as we can come up with a way of describing that policy in an understandable way.

We could combine the arguments for `snap-schedule set` and `prune-schedule set`, in other words tie the prune schedule spec directly to a snapshot schedule but keep the explicit `--keep-*` arguments in order to keep things understandable.

I can alter the interface accordingly and there would be some UI workflows to iron out (like how to set/change a pruning schedule without changing the snapshot schedule) but I don't anticipate any show stoppers.

How does that sound?

liewegas commented:
Something about the ``--keep-*`` args being explicitly tied to fixed periods (day, week, month) bugs me, but I'm not sure many/any users would need something different.

jan--f commented:
> Something about the `--keep-*` args being explicitly tied to fixed periods (day, week, month) bugs me, but I'm not sure many/any users would need something different.

Consider it the trade-off for a complex specification string. Imho this is pretty accessible if a bit verbose. Glancing over the options should be enough to figure out whats going on.

jan--f commented:
I updated the draft accordingly. One thing that is missing in this variant is how to alter pruning schedules explicitly. Not sure whether a separate subcommand or re-using `set` is better, though that should be easy enough to hash out later.

oliveiradan commented:
@jan--f , I actually was wondering if something like this: https://github.com/oliveiradan/ceph/commit/6ef977c67c2e52a826e975f3fcf21592a0228c22 
Would be interesting, either for a .json file, or even if later on using the ''Rados". The commented json at the top gives an idea.

jan--f commented:
I added the interface implementation. I changed the `set` interface somewhat to accommodate for the lack of kwarg passing to mgr modules. The retention policy is for now just a string. I think passing something like `keep-weekly=10,keep-monthly=4` would be good to have (the module would parse that accordingly) but we could also come up with something more succinct.

Some open questions:

- [ ] Should allow setting schedules on not-yet-exsiting paths?
- [ ] Should we allow setting a schedule without a retention policy?

 and TODOs:

- [ ] Remove two instances of recursive calls in the `ScheduleIndex`
- [ ] Add more tests
- [ ]  Add logging
- [ ] Improve output formatting

liewegas commented:
> I added the interface implementation. I changed the `set` interface somewhat to accommodate for the lack of kwarg passing to mgr modules. The retention policy is for now just a string. I think passing something like `keep-weekly=10,keep-monthly=4` would be good to have (the module would parse that accordingly) but we could also come up with something more succinct.
> 
> Some open questions:
> 
> - [ ] Should allow setting schedules on not-yet-exsiting paths?
> - [ ] Should we allow setting a schedule without a retention policy?

Would it simplify things to say no to both?

jan--f commented:
cc @batrick @vshankar

jan--f commented:
> Would it simplify things to say no to both?
> - [ ] Should allow setting schedules on not-yet-exsiting paths? 
Not allowing this for now is probably fine. I could see this as a feature request at some point though then probably more complex. E.g. the ability to add a schedule for a path glob to set a schedule for all dirs under /home/. But we can certainly cross that bridge when we come to it.

- [ ] Should we allow setting a schedule without a retention policy?
Allowing no retention policy is probably harder,since I assume we don't really want a large number of snapshots accumulating. So we'd need some kind of health check or so. But this is probably easier solved by setting a default policy (say keep the last 15 snapshots).

jan--f commented:
For retention specification the plan is currently to go with a string of `$n$timeperiod` concatenations. For example: `24h7d4w` meaning "keep the latest hourly snapshots for the past 24 hours, the latest daily snapshots for the past 7 days and the latest weekly snapshots for the last 4 weeks".

For $timeperiod I'd allow the obvious ones `h, d, w, m, y` and maybe `n` to specify to keep the last $n snapshots regardless of time.

vshankar commented on src/pybind/mgr/snap_schedule/module.py:
```diff
@@ -0,0 +1,124 @@
+            self.client.store_snap_schedule(use_fs, sched)
+            suc_msg = f'Schedule set for path {path}'
+        except sqlite3.IntegrityError:
+            existing_sched = self.client.get_snap_schedule(use_fs, path)
```
where is `get_snap_schedule`? typo probably?

jan--f commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,297 @@
+    def create_scheduled_snapshot(self, fs_info, path, retention):
+        self.log.debug(f'Scheduled snapshot of {path} triggered')
+        try:
+            time = datetime.datetime.utcnow().strftime('%Y-%m-%d-%H_%M_%S')
```
I think only handling utc internally makes sense. What's missing here is to handle conversion from a user supplied timezone.

jan--f replied:
Sorry my head was somewhere else...you're right of course. Though the timezone issue is still there.

jan--f commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,297 @@
+            INNER JOIN schedules_meta sm ON sm.schedule_id = s.id
+        ORDER BY until;'''
+
+    def refresh_snap_timers(self, fs):
```
hmm this is likely due to a compile time default of the sqlite library. I'll look into how to best resolve this.

jan--f commented:
restest this please

vshankar commented on src/pybind/mgr/snap_schedule/module.py:
```diff
@@ -0,0 +1,134 @@
+            report = [s.report() for s in existing_scheds]
+            error_msg = f'Found existing schedule {report}'
+            self.log.error(error_msg)
+            return errno.EEXISTS, '', error_msg
```
EEXIST

vshankar commented:
@jan--f pingback when this is ready for review

jan--f commented:
ping @vshankar I tihnk this is ready for a first review round. Still some things to discuss/improve but the functionality is all there.
Tests are next before some more refactoring.

jan--f commented:
@vshankar fd5321d is missing the Signed-off line. May I assume you're ok with me adding it in your name?

vshankar commented:
> @vshankar [fd5321d](https://github.com/ceph/ceph/commit/fd5321de3f52347a78cc2f7a21f2685fce28caab) is missing the Signed-off line. May I assume you're ok with me adding it in your name?

yeh please, thx

sebastian-philipp commented on src/pybind/mgr/snap_schedule/tox.ini:
```diff
@@ -0,0 +1,19 @@
+[tox]
+envlist = py27,py3
+skipsdist = true
+; toxworkdir = {env:CEPH_BUILD_DIR}/snap-schedule
+; minversion = 2.8.1
+
+[testenv]
+setenv=
+    LD_LIBRARY_PATH = {toxinidir}/../../../../build/lib
+    PATH = {toxinidir}/../../../../build/bin:$PATH
+    py27: PYTHONPATH = {toxinidir}/../../../../build/lib/cython_modules/lib.2:..
+    py3:  PYTHONPATH = {toxinidir}/../../../../build/lib/cython_modules/lib.3:{toxinidir}
+    CEPH_SNAP_SCHEDULE_UNITTEST = true
+deps =
+    pytest
+    mock
+    py27: pathlib
+commands=
+    pytest {posargs}
```
No need to add this. just add yourself to 

https://github.com/ceph/ceph/blob/90699f6e48c322ec8612939617e26df6305b8da1/src/pybind/mgr/tox.ini#L14-L22

sebastian-philipp commented on src/pybind/mgr/snap_schedule/__init__.py:
```diff
@@ -0,0 +1,6 @@
+
+import os
+
+if 'CEPH_SNAP_SCHEDULE_UNITTEST' not in os.environ:
+    from .module import Module
```
Would be great to take https://github.com/ceph/ceph/blob/master/src/pybind/mgr/insights/__init__.py as an example.

vshankar commented on src/pybind/mgr/snap_schedule/module.py:
```diff
@@ -0,0 +1,178 @@
+            return e.to_tuple()
+        if not scheds:
+            return errno.ENOENT, '', f'SnapSchedule for {path} not found'
+        return 0, '\n'.join([str(sched) for sched in scheds]), ''
```
Any reason this is not a json format like it was earlier? (the tests rely on output to be json)

jan--f replied:
I anticipate the `list` command being used by a syncing tool, to find out where to check for new snapshots to be synced. So the idea is to have output here that can be easily parsed with bash tools.

Probably good the add a `--format json` option.

vshankar commented on src/pybind/mgr/snap_schedule/module.py:
```diff
@@ -0,0 +1,178 @@
+            ret_scheds = self.client.get_snap_schedules(use_fs, path)
+        except CephfsConnectionException as e:
+            return e.to_tuple()
+        return 0, '\n===\n'.join([ret_sched.report() for ret_sched in ret_scheds]), ''
```
ditto as per list: any reason why this is not a json out?

jan--f replied:
ditto `--format json`

jan--f commented on src/pybind/mgr/snap_schedule/fs/schedule_client.py:
```diff
@@ -0,0 +1,261 @@
+SNAP_DB_PREFIX = 'snap_db'
+# increment this every time the db schema changes and provide upgrade code
+SNAP_DB_VERSION = '0'
+SNAP_DB_OBJECT_NAME = f'{SNAP_DB_PREFIX}_v{SNAP_DB_VERSION}'
```
The idea is to add that once we need change the schema and increment that version counter.
Once that happens, the db will be stored under a new object name, and the migration code (that would need to be added) would be responsible for migrating the v0 DB to v1 DB.

batrick commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,311 @@
+        Values(?, ?, ?, ?);'''
+    INSERT_SCHEDULE_META = '''INSERT INTO
+        schedules_meta(schedule_id, start, created, repeat, schedule, retention)
+        SELECT ?, ?, ?, ?, ?, ?'''
```
drop `created`

then

`SELECT ?, datetime('now'), ...`

batrick commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,311 @@
+
+    UPDATE_LAST = '''UPDATE schedules_meta
+    SET
+      last = ?,
```
this can also be `datetime('now')`

jan--f replied:
I was trying to keep the timestamp creation in the python realm. Motivation here is to have a reliable format and python just seems to handle this better.
Yes we can make sqlite fill in some default dates but this opens up hard to debug room for bugs. Consider the following for example on my laptop (UTC+2):
```
>>> import datetime
>>> datetime.datetime.now().isoformat()
'2020-05-19T15:42:04.905583'
>>> datetime.datetime.now(datetime.timezone.utc).isoformat()
'2020-05-19T13:42:07.115082+00:00'
>>> import sqlite3
>>> con = sqlite3.connect(":memory:")
>>> con.execute('select strftime("%Y-%m-%dT%H:%M.%S","now")').fetchall()[0]
('2020-05-19T13:42.27',)
>>> con.execute('select strftime("%Y-%m-%dT%H:%M.%S","now", "utc")').fetchall()[0]
('2020-05-19T11:42.37',)
```

So I would propose to simply keep timestamp creation in python land, for sanity's sake.

batrick commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,311 @@
+        return f'''{self.path} {self.schedule} {self.retention}'''
+
+    CREATE_TABLES = '''CREATE TABLE schedules(
+        id integer PRIMARY KEY ASC,
```
I know it doesn't matter but you use `INT`, `integer`, and `int` interchangeably. I'd prefer just using `INT`. Same for the other types below...

batrick commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,311 @@
+        start TEXT NOT NULL,
+        first TEXT,
+        last TEXT,
+        last_pruned TEXT,
```
`TIMESTAMP`

jan--f commented on src/pybind/mgr/snap_schedule/fs/schedule.py:
```diff
@@ -0,0 +1,311 @@
+    CREATE TABLE schedules_meta(
+        id INTEGER PRIMARY KEY ASC,
+        schedule_id INT,
+        start TEXT NOT NULL,
```
The reason I went with TEXT and iso datetime strings in UTC is that

1. sqlite doesn't know any date types. (see https://www.sqlite.org/datatype3.html)
2. No issues at all with with the whole 2038 problem.

The timestamp math that is used to calculate the next snapshot is the only part where unix epoch time makes sense (and that's what is used too). But otherwise it seems to be just as convenient to store timestamp strings.

batrick replied:
Right, but using `TIMESTAMP`, the raw type is actually TEXT. You can confirm:

````
$ python3
>>> import sqlite3
>>> import datetime
>>> db = sqlite3.connect('foo.db', detect_types=sqlite3.PARSE_DECLTYPES)
>>> c = db.execute('create table foo (bar integer, baz timestamp);')
>>> c=db.execute('insert into foo values(?, ?);', (2,datetime.datetime.now(),))
>>> c=db.execute('insert into foo values(?, ?);', (2,datetime.datetime.now(),))
>>> db.flush()
^D
$ sqlite3 foo.db
sqlite> select *, typeof(baz) from foo;
2|2020-05-19 09:07:16.595845|text
2|2020-05-19 09:08:45.850235|text
````

But, now that I look at this again. I see that the python3 adapter is inserting local times into the database. (`/headdesk`)

I was trying to be clever by suggesting the use of the python3 adapter which, now that I see it uses local time, really is utter shit. What I've done in the past for personal project is do:

`foo DATETIME NOT NULL DEFAULT (strftime('%s', 'now'))`

The `DATETIME` raw type is a integer for on-disk compactness. If you do basic dumps of the database it won't be readable as, obviously, it's a UNIX epoch. You can use strftime in your `SELECT `s to get a useful ISO string to pass to `datetime` (or just pass the UNIX epoch to datetime which is equiv.).

Anyway, this is all a suggestion.

sebastian-philipp commented on src/pybind/mgr/snap_schedule/tox.ini:
```diff
@@ -0,0 +1,19 @@
+[tox]
+envlist = py3
+skipsdist = true
+; toxworkdir = {env:CEPH_BUILD_DIR}/snap-schedule
+; minversion = 2.8.1
+
+[testenv]
+setenv=
+    LD_LIBRARY_PATH = {toxinidir}/../../../../build/lib
+    PATH = {toxinidir}/../../../../build/bin:$PATH
+    py27: PYTHONPATH = {toxinidir}/../../../../build/lib/cython_modules/lib.2:..
+    py3:  PYTHONPATH = {toxinidir}/../../../../build/lib/cython_modules/lib.3:{toxinidir}
+    CEPH_SNAP_SCHEDULE_UNITTEST = true
+deps =
+    pytest
+    mock
+    py27: pathlib
+commands=
+    pytest {posargs}
```
it still would be great to have your tests part of the mgr's tox file:

* It would reduce the amount of configs (tox.ini, cmake) to maintain
* It provides more visibility to everyone working on mgr modules
* `make check` runs faster due to not having to build another tox environment
* I already removed `insights/tox`, `orchestrator/tox.ini`, `ansible/tox.ini`.

jan--f commented:
@vshankar ready for a review round

jan--f commented:
Note for the retention interface:
It can be used similarly to the old interface: `ceph fs snap-schedule retention add /bar 12h7d4w`
or like discussed in #35220 :
```
ceph fs snap-schedule retention add /bar h 12
ceph fs snap-schedule retention add /bar d 7
ceph fs snap-schedule retention add /bar w 4
```

jan--f commented:
@vshankar anything blocking a review?

vshankar commented:
> @vshankar anything blocking a review?

I guess nothing -- although it has been a while since I've looked at this PR. Let's catch up tomorrow on weekly call and plan to merge this ASAP.

vshankar commented:
PR looks ok. I will go through it once more and ack.

vshankar commented:
jenkins retest this please

vshankar commented:
retest this please

jan--f commented:
rebased on master

vshankar commented:
retest this please

vshankar commented:
@jan--f please look at make check errors:

```
    def call(self, mgr, cmd_dict, inbuf):
        kwargs = {}
        kwargs_switch = False
        for a, d in self.args_dict.items():
            if 'req' in d and d['req'] == "false" and a not in cmd_dict:
                continue
>           if kwargs_switch or '=' in cmd_dict[a]:
E           TypeError: argument of type 'NoneType' is not iterable
```

jan--f commented:
Sigh...`datetime.fromisoformat` was add in python3.7. I'll explore working around that for python3.6

jan--f commented:
So seems like the ideal path is to package https://github.com/movermeyer/backports.datetime_fromisoformat and depend on that.

I could also rip out the use of `fromisoformat` but that would mean that users had to adhere to one strict know data string format to specify start times. Something like `%Y-%m-%d %H:%M%z` -> `2020-08-07 15:15+02:00`

What do you think @vshankar @batrick ?

batrick commented:
> So seems like the ideal path is to package https://github.com/movermeyer/backports.datetime_fromisoformat and depend on that.
> 
> I could also rip out the use of `fromisoformat` but that would mean that users had to adhere to one strict know data string format to specify start times. Something like `%Y-%m-%d %H:%M%z` -> `2020-08-07 15:15+02:00`
> 
> What do you think @vshankar @batrick ?

I would suggest you revisit my suggestion here: https://github.com/ceph/ceph/pull/29489#discussion_r427431288

Store UNIX epochs and convert to datetime in python land. I don't think it's valuable storing sub-second datetimes in sqlite. I also don't suggest getting fancy with automatic date conversions between sqlite<->python land using the sqlite python adapters. They're complete crap as I noted in my comment above. What you're already doing here

https://github.com/ceph/ceph/pull/29489/files#diff-a6deb637f7e7db8b4b21ecdf0952a01fR97-R110

should be sufficient if you create the datetime from the unix epoch.

batrick commented:
@jan--f one more thing, we don't have any docs for this work. You can either add them to this PR while you fix up the datetime issues or please setup a tracker ticket so it's not forgotten.

jan--f commented:
> I would suggest you revisit my suggestion here: [#29489 (comment)](https://github.com/ceph/ceph/pull/29489#discussion_r427431288)
> 
> Store UNIX epochs and convert to datetime in python land. I don't think it's valuable storing sub-second datetimes in sqlite. I also don't suggest getting fancy with automatic date conversions between sqlite<->python land using the sqlite python adapters. They're complete crap as I noted in my comment above. What you're already doing here
> 
> https://github.com/ceph/ceph/pull/29489/files#diff-a6deb637f7e7db8b4b21ecdf0952a01fR97-R110
> 
> should be sufficient if you create the datetime from the unix epoch.

This is not an issue of internal representation. Internally I can (and will) simply define a date format and parse date strings via `strptime`. I'd like to avoid storing unix epochs, since there is a good chance that snapshot schedules will exist a long time. Given that there is no downside to storing date strings (in a predefined format) compared to epochs, but time strings skirt the Year 2038 problem neatly I opted for this.

The question raised here is one of user interface. Without python3.7's `fromisoformat` I'll have to require user specified dates in a predefined format. I.e. a start time of, say midnight tonight would have to be passed as `2020-08-10T00:00+02:00` instead of just `00:00`.

I think I'll make the iso date parsing optional. Either a user runs py3.7+, they install `backports.datetime_fromisoformat` or they have to deal with a longer data format.

jan--f commented:
> @jan--f one more thing, we don't have any docs for this work. You can either add them to this PR while you fix up the datetime issues or please setup a tracker ticket so it's not forgotten.

Docs are being worked on.

jan--f commented:
current make check failure seems unrelated.

jan--f commented:
@vshankar ready for another round.

vshankar commented:
> @vshankar ready for another round.

ack -- will take a look.

vshankar commented:
retest this please

sebastian-philipp commented on src/pybind/mgr/snap_schedule/__init__.py:
```diff
@@ -0,0 +1,11 @@
+
+from os import environ
+
+if 'SNAP_SCHED_UNITTEST' in environ:
+    import tests
```
(just fyi, you can run only your test via 

```
cd src/pybind/mgr
../../script/run_tox.sh --tox-envs=py3,mypy mgr -- snap_schedule
```
)

vshankar commented:
LGTM -- I'm running this with PR #34552 (not update yet).

make check failure looks unrelated -- but required to merge this.

batrick commented:
jenkins test make check

vshankar commented:
jenkins test make check

vshankar commented:
Test run with this PR and #34552 (w/ some formatting fixes): https://pulpito.ceph.com/vshankar-2020-08-20_11:09:05-fs-wip-cephfs-snap-schedule-20-08-20-054521-distro-basic-smithi/

`Test failure: test_multi_snap_schedule (tasks.cephfs.test_snap_schedules.TestSnapSchedules)`

From manager logs:

```
2020-08-20T11:25:00.962+0000 7f29624e3700  0 [snap_schedule INFO snap_schedule.fs.schedule_client] created scheduled snapshot of snap_test_dir1
2020-08-20T11:25:00.966+0000 7f29624e3700  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] rmdir on scheduled-2020-08-20-11_25_00
2020-08-20T11:26:00.970+0000 7f295e4db700  0 [snap_schedule INFO snap_schedule.fs.schedule_client] created scheduled snapshot of snap_test_dir1
2020-08-20T11:26:00.972+0000 7f295e4db700  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] rmdir on scheduled-2020-08-20-11_26_00
```

The test (`test_multi_snap_schedule`) adds two schedules for a path and expects the snapshots to be present after the elapsed time. Looks that the snapshots got pruned as soon as it got created. So the test could not find them.

@jan--f any idea? This does not happen on every run -- looks like some kind of race?

vshankar commented:
One more thing: I was using minute granularity for tests if it matters. Since it does not happen all the time -- I think we should merge this PR and open a tracker ticket for the failed test. @jan--f @batrick?

The tox test (run-tox-mgr) is failing consistently. @jan--f could you please take a look at that?

batrick commented:
Let's fix the tox errors then merge this. Then open tickets for outstanding bugs.

jan--f commented:
jenkins test make check

jan--f commented:
The make check issue seems like a cephadm issue. I asked @sebastian-philipp to have a look.

sebastian-philipp commented:
https://tracker.ceph.com/issues/47141

sebastian-philipp commented:
Hm. Indeed. this fixes the issue:

```diff
diff --git a/src/pybind/mgr/snap_schedule/__init__.py b/src/pybind/mgr/snap_schedule/__init__.py
index 8001b71848..04314485ed 100644
--- a/src/pybind/mgr/snap_schedule/__init__.py
+++ b/src/pybind/mgr/snap_schedule/__init__.py
@@ -6,6 +6,7 @@ if 'SNAP_SCHED_UNITTEST' in environ:
     import tests
 elif 'UNITTEST' in environ:
     import tests
+    tests.mock_ceph_modules()  # type: ignore
     from .module import Module
 else:
     from .module import Module
```

Looks like we have to decide between:

1. Mocking the `rados` module
2. Importing the `rados` module without using it

jan--f commented:
Mocking `rados` seems reasonable to me. The snap-schedule unit tests don't need it (yet).

jan--f commented:
> @jan--f any idea? This does not happen on every run -- looks like some kind of race?

@vshankar  Yeah an empty retention policy was not treated correctly. Fixed now, if you want to re-run the test.

vshankar commented:
> @vshankar Yeah an empty retention policy was not treated correctly. Fixed now, if you want to re-run the test.

I'll vote to merge this (once make check passes) and create a tracker for the issue

jan--f commented:
fine with me. make check is fixed in #36811

tchaikov commented:
@vshankar how did you test this change?

vshankar commented:
> @vshankar how did you test this change? my batch now looks like https://pulpito.ceph.com/kchai-2020-08-29_13:54:19-rados-wip-kefu-testing-2020-08-29-1945-distro-basic-smithi/

with this pr: https://github.com/ceph/ceph/pull/34552 + tox test that run on PR updates. Are you suspecting this part of the change to be the issue: https://github.com/ceph/ceph/pull/29489/files#diff-af4be246b95af4d67b3fc62e30a49178

tchaikov commented:
> > @vshankar how did you test this change? my batch now looks like https://pulpito.ceph.com/kchai-2020-08-29_13:54:19-rados-wip-kefu-testing-2020-08-29-1945-distro-basic-smithi/
> 
> with this pr: #34552 + tox test that run on PR updates. Are you suspecting this part of the change to be the issue: https://github.com/ceph/ceph/pull/29489/files#diff-af4be246b95af4d67b3fc62e30a49178

@vshankar i don't think tox based tests have good coverage for these changes or they are able to find most the issues. in future, i'd suggest run any non-trivial change with rados suite. see my analysis at https://tracker.ceph.com/issues/47185

tchaikov commented:
@vshankar please add a Reviewed-by line in your merge commit when merge a PR. in future

vshankar commented:
> @vshankar please add a Reviewed-by line in your merge commit when merging a PR in future.

ACK.
